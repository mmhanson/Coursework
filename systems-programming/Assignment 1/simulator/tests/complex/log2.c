#include <stdio.h>

// Recursively computes log base 2
// of the input, by counting how many
// times we can divide it by 2
unsigned log2(unsigned input)
{

  // Integer math only, approximation
  if(input <= 2)
  {
    return 1;
  }

  return 1 + log2(input / 2);
}

int main()
{
  int input;
  scanf("%d", &input);
  return log2(input);
}
