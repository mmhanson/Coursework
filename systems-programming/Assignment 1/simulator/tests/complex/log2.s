main:
	subl	$8, %esp
	readr	%edi
	call	log2
	printr	%eax
	addl	$8, %esp
	ret
log2:
	movl	$2, %r8d
	cmpl	%r8d, %edi
	jbe	.L3
	subl	$8, %esp
	shrl	%edi
	call	log2
	addl	$1, %eax
	jmp	.L2
.L3:
	movl	$1, %eax
	ret
.L2:
	addl	$8, %esp
	ret
