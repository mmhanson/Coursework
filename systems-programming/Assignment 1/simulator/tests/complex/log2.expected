4 (0x4)
